catenets/models/torch/snet.py

```diff
@@ -0,0 +1,342 @@
+
+            if i % self.n_iter_print == 0:
+                with torch.no_grad():
+                    po_preds, prop_preds = self._step(X_val)
+                    val_loss = self.loss(po_preds, prop_preds, y_val, w_val)
+                    log.info(
+                        f"[{self.name}] Epoch: {i}, current {val_string} loss: {val_loss} train_loss: {torch.mean(train_loss)}"
+                    )
+
+        return self
+
+    @abc.abstractmethod
+    def _step(self, X: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]:
+        ...
+
+    def _forward(self, X: torch.Tensor) -> torch.Tensor:
+        X = torch.Tensor(X)
+        repr_preds = self._repr_estimator(X).squeeze()
+        y0_preds = self._po_estimators[0](repr_preds).squeeze()
+        y1_preds = self._po_estimators[1](repr_preds).squeeze()
+
+        return torch.vstack((y0_preds, y1_preds)).T
+
+    def forward(self, X: torch.Tensor) -> torch.Tensor:
+        """
+        Predict treatment effects and potential outcomes
+
+        Parameters
+        ----------
+        X: array-like of shape (n_samples, n_features)
+            Test-sample features
+        Returns
+        -------
+        y: array-like of shape (n_samples,)
+        """
+        preds = self._forward(X)
+        y0_preds = preds[:, 0]
+        y1_preds = preds[:, 1]
+
+        return y1_preds - y0_preds
+
+
+class TARNet(SNet):
+    def __init__(
+        self,
+        n_unit_in: int,
+        binary_y: bool = False,
+        n_layers_out: int = DEFAULT_LAYERS_OUT,
+        n_units_out: int = DEFAULT_UNITS_OUT,
+        nonlin: str = DEFAULT_NONLIN,
+        **kwargs: Any,
+    ) -> None:
+        propensity_estimator = PropensityNet(
+            "tarnet_propensity_estimator",
+            n_unit_in,
+            2,
+            "prop",
+            n_layers_out_prop=n_layers_out,
+            n_units_out_prop=n_units_out,
+            nonlin=nonlin,
+        ).to(DEVICE)
+        super(TARNet, self).__init__(
+            "TARNet",
+            n_unit_in,
+            propensity_estimator,
+            binary_y=binary_y,
+            n_layers_out=n_layers_out,
+            n_units_out=n_units_out,
+            nonlin=nonlin,
+            **kwargs,
+        )
+
+    def _step(self, X: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]:
+        po_preds = self._forward(X)
+        prop_preds = self._propensity_estimator(X)
+
+        return po_preds, prop_preds
+
+
+class DragonNet(SNet):
+    def __init__(
+        self,
+        n_unit_in: int,
+        binary_y: bool = False,
+        n_layers_out: int = DEFAULT_LAYERS_OUT,
+        n_units_out: int = DEFAULT_UNITS_OUT,
+        nonlin: str = DEFAULT_NONLIN,
+        n_units_r: int = DEFAULT_UNITS_R_BIG_S,
+        **kwargs: Any,
+    ) -> None:
+        propensity_estimator = PropensityNet(
+            "dragonnet_propensity_estimator",
+            n_units_r,
+            2,
+            "prop",
+            n_layers_out_prop=n_layers_out,
```

Comment:
for DragonNet, n_layers_out_prop and n_units_out_prop should definitely be allowed to be different from n_layers_out -- in the original DragonNet paper, n_layers_out_prop=0 (so that should be the default here).